var/spack/repos/builtin/packages/ghostscript/package.py

```diff
@@ -36,6 +38,10 @@ class Ghostscript(AutotoolsPackage):
 
     patch('nogoto.patch', when='%fj@:4.1.0')
 
+    # Related bug report: https://bugs.ghostscript.com/show_bug.cgi?id=702985
+    # Patch source: https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff;h=41ef9a0bc36b
+    patch("https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b", when='@9.53.3^freetype@2.10.3:', sha256="254532a8561afddb5b39c1c30526676559ed9f0b87c1e9985eb797a59c98a3dd")
```

Comment:
The patch is not retrievable from that url. Maybe use a github mirror of the ghostscript git repository since github can generate retrievable patches.

Reply:
Interesting... It's fetching just fine for me

Reply:
Can you try running `spack clean --all` and verify that it still fetches?

Here is what I get:
```
$ spack patch ghostscript@:9.53
==> Fetching https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs9533/ghostscript-9.53.3.tar.gz
######################################################################## 100.0%
==> Fetching file:///nfs/install/spack/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Error: [Errno 2] No such file or directory: '/tmp/gpjohnsn/spack-stage/spack-stage-7wbz9cav/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b.part' -> '/tmp/gpjohnsn/spack-stage/spack-stage-7wbz9cav/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b'
```

This url works though:
`https://github.com/ArtifexSoftware/ghostpdl/commit/41ef9a0bc36b9db7115fbe9623f989bfb47bbade.patch`

The sha256 would also need to be changed.
`sha256='49c353106d97c40b3b2c78f72ce34e3eef66e6b04861c313f87bad11ab4189e6'

Reply:
```
jchits@sporcbuild ~/workspace/spack/var/spack/repos on jchits/rit/ghostscript(v0.16.0-1802-g9aaa028ac7)
repos:jchits/rit/ghostscript % spack clean --all                                                                                                                   2021-04-19 15:14:59 jchits pts/3
==> Removing all temporary build stages
==> Removing cached downloads
==> Removing install failure marks
==> Removing cached information on repositories
==> Removing python cache files
jchits@sporcbuild ~/workspace/spack/var/spack/repos on jchits/rit/ghostscript(v0.16.0-1802-g9aaa028ac7)
repos:jchits/rit/ghostscript % spack patch ghostscript@:9.53                                                                                                       2021-04-19 16:01:13 jchits pts/3
==> Fetching https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs9533/ghostscript-9.53.3.tar.gz
######################################################################## 100.0%
==> Fetching https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b
######################################################################## 100.0%
```

Reply:
This looks like a problem with the interaction of local spack mirrors and the characters in the url string of the patch.
If I remove my local spack mirror then the download works fine. The patch is not actually in the local spack mirror because adding the patch to the mirror also fails. Using the ghostscript github mirror as a source for the patch works fine.

I think the best course of action is to use the patch from the github mirrored source of ghostscript, which is where the tarfile is from as well. So,
```python
patch('https://github.com/ArtifexSoftware/ghostpdl/commit/41ef9a0bc36b9db7115fbe9623f989bfb47bbade.patch',                              
          sha256='49c353106d97c40b3b2c78f72ce34e3eef66e6b04861c313f87bad11ab4189e6',                                                        
          when='^freetype@2.10.3:')
```

Reply:
@alalazo This probably needs an issue filed, but I will capture it here for now.
```console
$ spack mirror list
local           file:///opt/packages/gpjohnsn/argon_mirror
spack-public    https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/
```
```console
$ spack mirror --destroy --mirror-name local
```
```console
$ spack mirror create --directory /opt/packages/gpjohnsn/argon_mirror/ ghostscript                             
==> Adding package ghostscript@9.50 to mirror
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Warning: Error while fetching ghostscript@9.50
  [Errno 2] No such file or directory: '/tmp/gpjohnsn/spack-stage/spack-stage-0dvv_qm3/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b.part' -> '/tmp/gpjohnsn/spack-stage/spack-stage-0dvv_qm3/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b'
==> Successfully updated mirror in file:///opt/packages/gpjohnsn/argon_mirror
  Archive stats:
    1    already present
    0    added
    1    failed to fetch.
==> Error: Failed downloads:
ghostscript@9.50
```
```console
$ spack patch ghostscript
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Error: [Errno 2] No such file or directory: '/tmp/gpjohnsn/spack-stage/spack-stage-iz8gs53_/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b.part' -> '/tmp/gpjohnsn/spack-stage/spack-stage-iz8gs53_/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b'
```
```console
$ spack mirror rm local
==> Removed mirror local.
```
```console
$ spack patch ghostscript
==> Using cached archive: /opt/packages/gpjohnsn/spack/var/spack/cache/_source-cache/archive/0f/0f53e89fd647815828fc5171613e860e8535b68f7afbc91bf89aee886769ce89.tar.gz
==> Fetching https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b
######################################################################## 100.0%
==> Applied patch https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b
==> Ran patch() for ghostscript
```
```console
$ spack mirror add local /opt/packages/gpjohnsn/argon_mirror/
```
```console
$ spack clean --all
==> Removing all temporary build stages
==> Removing cached downloads
==> Removing install failure marks
==> Removing cached information on repositories
==> Removing python cache files
==> Removing software in "/Users/gpjohnsn/.spack/bootstrap/store"
```
```console
$ spack patch ghostscript
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/_source-cache/archive/0f/0f53e89fd647815828fc5171613e860e8535b68f7afbc91bf89aee886769ce89.tar.gz
######################################################################## 100.0%
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/ghostscript/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b-254532a

==> Error: [Errno 2] No such file or directory: '/tmp/gpjohnsn/spack-stage/spack-stage-olc7wfht/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b.part' -> '/tmp/gpjohnsn/spack-stage/spack-stage-olc7wfht/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b'
```
I can add to the mirror if the mirror is not enabled
```console
$ spack mirror rm local                                                                                       
==> Removed mirror local.
...
$ spack mirror create --directory /opt/packages/gpjohnsn/argon_mirror/  ghostscript
==> Adding package ghostscript@9.50 to mirror
==> Fetching https://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff_plain;h=41ef9a0bc36b
######################################################################## 100.0%
==> Successfully updated mirror in file:///opt/packages/gpjohnsn/argon_mirror
  Archive stats:
    1    already present
    1    added
    0    failed to fetch.
```
Everything works as expected with the patch from github.
```console
$ spack mirror destroy --mirror-name local
...
$ spack mirror list
local           file:///opt/packages/gpjohnsn/argon_mirror
spack-public    https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/
...
$ spack mirror create --directory /opt/packages/gpjohnsn/argon_mirror/  ghostscript                             
==> Adding package ghostscript@9.50 to mirror
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/0f/0f53e89fd647815828fc5171613e860e8535b68f7afbc91bf89aee886769ce89.tar.gz
######################################################################## 100.0%
==> Fetching https://github.com/ArtifexSoftware/ghostpdl/commit/41ef9a0bc36b9db7115fbe9623f989bfb47bbade.patch
######################################################################## 100.0%
==> Successfully created mirror in file:///opt/packages/gpjohnsn/argon_mirror
  Archive stats:
    0    already present
    1    added
    0    failed to fetch.
...
$ spack patch ghostscript
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/_source-cache/archive/0f/0f53e89fd647815828fc5171613e860e8535b68f7afbc91bf89aee886769ce89.tar.gz
######################################################################## 100.0%
==> Fetching file:///opt/packages/gpjohnsn/argon_mirror/_source-cache/archive/49/49c353106d97c40b3b2c78f72ce34e3eef66e6b04861c313f87bad11ab4189e6
######################################################################## 100.0%
==> Applied patch https://github.com/ArtifexSoftware/ghostpdl/commit/41ef9a0bc36b9db7115fbe9623f989bfb47bbade.patch
==> Ran patch() for ghostscript
```
To reiterate, having an active mirror that does not have the patch, but does have the package directory in the local mirror, and the patch source has `?` and `=` characters, causes the patch stage to fail.

Reply:
@qwertos @glennpj Thanks for the report. I agree this should be tracked in its own issue. I'll open one later if time permits.